tt_metal/impl/device/device.cpp

```diff
@@ -3405,6 +3553,113 @@ void Device::generate_device_headers(const std::string &path) const
+    if (this->active_sub_device_manager_id_.has_value()) {
+        const auto& sub_device_manager = this->sub_device_managers_.at(*this->active_sub_device_manager_id_);
+        return sub_device_manager->noc_unicast_data(sub_device_id);
+    } else {
+        TT_FATAL(sub_device_id < Device::DEFAULT_NUM_SUB_DEVICES, "sub_device_id {} is out of range", sub_device_id);
+        return this->noc_unicast_data_;
+    }
+}
+
+const vector_memcpy_aligned<uint32_t>& Device::noc_mcast_unicast_data(uint32_t sub_device_id, bool mcast_data, bool unicast_data) const {
+    // Needed for compatibility with tests that create programs with no kernels
+    static const vector_memcpy_aligned<uint32_t> empty = {};
+    if (this->active_sub_device_manager_id_.has_value()) {
+        const auto& sub_device_manager = this->sub_device_managers_.at(*this->active_sub_device_manager_id_);
+        if (mcast_data && unicast_data) {
+            return sub_device_manager->noc_mcast_unicast_data(sub_device_id);
+        } else if (mcast_data) {
+            return sub_device_manager->noc_mcast_data(sub_device_id);
+        } else if (unicast_data) {
+            return sub_device_manager->noc_unicast_data(sub_device_id);
+        } else {
+            return empty;
+        }
+    } else {
+        TT_FATAL(sub_device_id < Device::DEFAULT_NUM_SUB_DEVICES, "sub_device_id {} is out of range", sub_device_id);
+        if (mcast_data && unicast_data) {
+            return this->noc_mcast_unicast_data_;
+        } else if (mcast_data) {
+            return this->noc_mcast_data_;
+        } else if (unicast_data) {
+            return this->noc_unicast_data_;
+        } else {
+            return empty;
+        }
+    }
+}
+
+uint32_t Device::num_noc_mcast_txns(uint32_t sub_device_id) const {
+    return this->noc_mcast_data(sub_device_id).size() / 2;
+}
+uint32_t Device::num_noc_unicast_txns(uint32_t sub_device_id) const {
+    return this->noc_unicast_data(sub_device_id).size();
+}
+
+uint32_t Device::num_noc_mcast_unicast_txns(uint32_t sub_device_id, bool mcast_data, bool unicast_data) const {
+    return (mcast_data ? this->num_noc_mcast_txns(sub_device_id) : 0) + (unicast_data ? this->num_noc_unicast_txns(sub_device_id) : 0);
+}
+
+NOC Device::dispatch_go_signal_noc() const {
+    return this->dispatch_s_enabled() ? NOC::NOC_1 : NOC::NOC_0;
+}
+
+std::optional<SubDeviceManagerId> Device::get_active_sub_device_manager_id() const {
+    return this->active_sub_device_manager_id_;
+}
+
+SubDeviceManagerId Device::create_sub_device_manager(tt::stl::Span<const SubDevice> sub_devices, DeviceAddr local_l1_size) {
+    static std::atomic<SubDeviceManagerId> sub_device_manager_id = 0;
+    TT_FATAL(!this->using_slow_dispatch(), "Using sub device managers is unsupported with slow dispatch");
+    auto [sub_device_manager, _] = this->sub_device_managers_.insert_or_assign(sub_device_manager_id, std::make_unique<detail::SubDeviceManager>(sub_devices, local_l1_size, this));
```

Comment:
Seeing as this appears to be the only time the `SubDeviceManager` is ever constructed, is it necessary to keep the constructor overload that move-constructs a vector? I don't see it being used anywhere.